src/scan_shadow_detector.cpp

```diff
@@ -54,22 +65,35 @@ namespace laser_filters
       max_angle_tan_ = -max_angle_tan_;
   }
 
-  bool ScanShadowDetector::isShadow(const float r1, const float r2, const float included_angle)
-  {
-    const float perpendicular_y_ = r2 * sinf(included_angle);
-    const float perpendicular_x_ = r1 - r2 * cosf(included_angle);
-    const float perpendicular_tan_ = fabs(perpendicular_y_) / perpendicular_x_;
+  void ScanShadowDetector::prepareForInput(const float angle_increment) {
```

Comment:
Having `prepareForInput()` be a separate function that the Filter class must remember to call seems like it could make it easy to add a bug later (someone doesn't know it has to be called). Why not call is every time in `ScanShadowDetector::isShadow()`?

Reply:
The drawback of moving it to the isShadow method is that it gets called (num_samples * window * 2) times per update instead of once. This results in about 5% regression in performance.

I propose to leave it as is. The risk of a regression is mitigated by the new unit test: SingleForwardShadow_AngleIncrementChanged. This test explicitly tests that the map is updated when needed and fails if this is not done.

Reply:
Ah, I see. Before this PR, the `ScanShadowDectector` class only does the computation for a given pair of laser beams - it doesn't care how many beams are in the laser scan or how they are spaced. After the PR, `ScanShadowDectector` makes an assumption (which is true for all LaserScan messages) that there is a fixed number of beams in a scan, and that they are evenly spaced. 

I tend to be very conservative about changes to the class interfaces in this package. The code has been around for a long time, and is likely used in many different ways by different people. For example, nothing stops people from using the `ScanShadowDetector` class directly from their own code without using the filter node. That isn't ideal, but I expect that it does happen. Whenever possible on code as old and stable as this, I'd prefer to make changes that break other use cases as little as possible.

What do you think about putting the cached sin and cos maps in the `ScanShadowFilter` class, since it already has a concept of a "scan" rather than just dealing with individual beams like the `ScanShadowDetector` class? Then you could add a lower level function to `ScanShadowDetector` that skips the computation of sin and cos:

```c++

bool ScanShadowDetector::isShadow2(float r1, float r2, float included_angle_sin, float included_angle_cos)
{
  const float perpendicular_y_ = r2 * included_angle_sin;
  const float perpendicular_x_ = r1 - r2 * included_angle_cos;
  const float perpendicular_tan_ = fabs(perpendicular_y_) / perpendicular_x_;
  
  if (perpendicular_tan_ > 0)
  {
    if (perpendicular_tan_ < min_angle_tan_)
      return true;
  }
  else
  {
    if (perpendicular_tan_ > max_angle_tan_)
      return true;
  }
}

bool ScanShadowDetector::isShadow(const float r1, const float r2, const float included_angle)
{
  float included_angle_sin = sinf(included_angle);
  float included_angle_cos = cosf(included_angle);
  return isShadow2(r1, r2, included_angle_sin, included_angle_cos)
}
```

Then the `ScanShadowFilter` class can handle caching the cos and sin values and call `ScanShadowDetector::isShadow2` to avoid doing the trig functions every time.

If i were writing this code from scrath, I would probably get rid of the ScanShadowDetector class completely and replace it with functions. Oh well.

Reply:
Okay, makes sense. We moved the look-up tables to the filter. The new isShadow method that takes the sin/cos values has become a private API in the detector, only for use by the filter.